src/app/tests/TestNullable.cpp

```diff
@@ -0,0 +1,307 @@
+        NL_TEST_ASSERT(inSuite, nullable2.SetToMatch(nullable1) == true);
+        NL_TEST_ASSERT(inSuite, nullable1 == nullable2);
+    }
+
+    // Structs
+    {
+        struct SomeObject
+        {
+            uint8_t a;
+            uint8_t b;
+
+            bool operator==(const SomeObject & other) const { return (a == other.a) && (b == other.b); }
+        };
+
+        auto nullable1 = MakeNullable<SomeObject>({ 1, 2 });
+        auto nullable2 = MakeNullable<SomeObject>({ 1, 2 });
+
+        NL_TEST_ASSERT(inSuite, !nullable1.IsNull());
+        NL_TEST_ASSERT(inSuite, !nullable2.IsNull());
+        NL_TEST_ASSERT(inSuite, nullable1 == nullable2);
+
+        // No-op on change to same.
+        NL_TEST_ASSERT(inSuite, nullable1.SetToMatch(nullable2) == false);
+        NL_TEST_ASSERT(inSuite, nullable1 == nullable2);
+
+        nullable1.Value().a = 100;
+
+        NL_TEST_ASSERT(inSuite, nullable1 != nullable2);
+        NL_TEST_ASSERT(inSuite, nullable2.SetToMatch(nullable1) == true);
+        NL_TEST_ASSERT(inSuite, nullable1 == nullable2);
+    }
+
+    // Scalar cases
+    {
+        auto nullable1 = MakeNullable(static_cast<uint8_t>(1));
+
+        NL_TEST_ASSERT(inSuite, !nullable1.IsNull());
+
+        // Non-null to non-null same value
+        NL_TEST_ASSERT(inSuite, nullable1.SetToMatch(nullable1) == false);
+        NL_TEST_ASSERT(inSuite, !nullable1.IsNull());
+
+        // Non-null to null
+        NL_TEST_ASSERT(inSuite, nullable1.SetToMatch(NullNullable) == true);
+        NL_TEST_ASSERT(inSuite, nullable1.IsNull());
+
+        // Null to null
+        NL_TEST_ASSERT(inSuite, nullable1.SetToMatch(NullNullable) == false);
+        NL_TEST_ASSERT(inSuite, nullable1.IsNull());
+
+        // Null to non-null
+        NL_TEST_ASSERT(inSuite, nullable1.SetToMatch(MakeNullable(static_cast<uint8_t>(1))) == true);
+        NL_TEST_ASSERT(inSuite, !nullable1.IsNull());
+        NL_TEST_ASSERT(inSuite, nullable1.Value() == 1);
+
+        // Non-null to non-null different value
+        NL_TEST_ASSERT(inSuite, nullable1.SetToMatch(MakeNullable(static_cast<uint8_t>(2))) == true);
+        NL_TEST_ASSERT(inSuite, !nullable1.IsNull());
+        NL_TEST_ASSERT(inSuite, nullable1.Value() == 2);
+
+        // Non-null to extent of range --> changes to "invalid" value in range.
+        NL_TEST_ASSERT(inSuite, nullable1.SetToMatch(MakeNullable(static_cast<uint8_t>(255))) == true);
+        NL_TEST_ASSERT(inSuite, !nullable1.IsNull());
+        NL_TEST_ASSERT(inSuite, nullable1.Value() == 255);
+    }
+}
+
+// clang-format off
+static const nlTest sTests[] =
+{
+    NL_TEST_DEF("NullableBasic", TestBasic),
+    NL_TEST_DEF("NullableMake", TestMake),
+    NL_TEST_DEF("NullableCopy", TestCopy),
+    NL_TEST_DEF("NullableMove", TestMove),
+    NL_TEST_DEF("Nullable SetToMatch operation", TestSetToMatch),
+    NL_TEST_SENTINEL()
+};
+// clang-format on
+
+int TestNullable_Setup(void * inContext)
+{
+    return SUCCESS;
```

Comment:
you can pass in nullptr to the nlTestSuite and not define these if they are a noop

Reply:
Done